File: 01_Python/02_LeetCode/0700_SearchInABinarySearchTree.py
"""
Problem description
 - https://leetcode.com/problems/search-in-a-binary-search-tree/

Result
 - Runtime: 72 ms, faster than 72.42% of Python online submissions for Search in a Binary Search Tree.
 - Memory Usage: 15 MB, less than 100.00% of Python online submissions for Search in a Binary Search Tree.
"""


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def searchBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        result = None

        if root.val == val:
            return root

        if val < root.val and root.left:
            result = self.searchBST(root.left, val)
        elif val > root.val and root.right:
            result = self.searchBST(root.right, val)

        return result

File: 02_JavaScript/02_LeetCode/0700_SearchInABinarySearchTree.js
/*
Problem description
  - https://leetcode.com/problems/search-in-a-binary-search-tree/

Result
  - Runtime: 68 ms, faster than 98.30% of JavaScript online submissions for Search in a Binary Search Tree.
  - Memory Usage: 41.8 MB, less than 56.25% of JavaScript online submissions for Search in a Binary Search Tree.
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
const searchBST = function (root, val) {
  if (!root) {
    return null;
  }

  if (root.val === val) {
    return root;
  }

  if (root.val > val) {
    return searchBST(root.left, val)
  }

  if (root.val < val) {
    return searchBST(root.right, val);
  }

  return null;
};